Bind ForEachVariableStatementSyntax.Variables for error recovery

333fred opened the pull request:
We now bind ForEachVariableStatementSyntax.Variables in error
scenarios, so that IOperation can provide a meaningful tree
in this scenario.

<details><summary>Ask Mode template not completed</summary>

<!-- This template is not always required. If you aren't sure about whether it's needed or want help filling out the sections,
submit the pull request and then ask us for help. :) -->

### Customer scenario

In an invalid code scenario where the user puts an arbitrary expression in a foreach block, we don't bind the expression in initial bind leaving information out of the IOperation tree.

### Bugs this fixes

Fixes https://github.com/dotnet/roslyn/issues/19996.
Fixes https://github.com/dotnet/roslyn/issues/23820.

### Workarounds, if any

None

### Risk

This will cause us to bind an additional expression during initial bind in foreach nodes, which could potentially be quite large. However, this will only affect a fairly-uncommon broken code scenario.

### Performance impact

Low, as the additional bind shouldn't adversely anything except the broken scenario.

### Is this a regression from a previous update?

No

### Root cause analysis

N/A

### How was the bug found?

Internally-reported

### Test documentation updated?

N/A

</details>

333fred commented on src/Compilers/CSharp/Test/Semantic/Semantics/OutVarTests.cs:
```diff
@@ -1073,14 +1073,6 @@ private static void AssertTypeFromOperation(SemanticModel model, TypeSymbol expe
             // unlike GetSymbolInfo or GetTypeInfo, GetOperation doesn't use SemanticModel's recovery mode.
             // what that means is that GetOperation might return null for ones GetSymbol/GetTypeInfo do return info from
             // error recovery mode
-            var foreachLoop = decl.Ancestors().OfType<ForEachVariableStatementSyntax>().FirstOrDefault();
```
This change causes `GetEnclosingBinderInternalRecovery_11` to validate IOperation nodes now.

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/ForEachLoopBinder.cs:
```diff
@@ -303,6 +304,12 @@ private BoundForEachStatement BindForEachPartsWorker(DiagnosticBag diagnostics,
+                        else
+                        {
+                            // Bind the expression for error recovery, but discard all new diagnostics
+                            iterationErrorExpression = originalBinder.BindExpression(node.Variable, new DiagnosticBag());
```
>iterationErrorExpression = originalBinder.BindExpression(node.Variable, new DiagnosticBag()); [](start = 28, length = 93)

It doesn't feel right that we are skipping previous ```else if (!node.HasErrors)```. I think the rule we should follow is simple, the expression should always be bound one way or another. Either we BindDeconstruction, or we BindExpression. There should be no way to skip binding.

AlekseyTs commented on src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs:
```diff
@@ -1457,6 +1457,10 @@ private IForEachLoopOperation CreateBoundForEachStatementOperation(BoundForEachS
                 var declaratorSyntax = boundForEachStatement.IterationVariableType.Syntax;
                 loopControlVariable = new Lazy<IOperation>(() => new VariableDeclarator(local, initializer: null, ignoredArguments: ImmutableArray<IOperation>.Empty, semanticModel: _semanticModel, syntax: declaratorSyntax, type: null, constantValue: default, isImplicit: false));
             }
+            else if (boundForEachStatement.IterationErrorExpressionOpt != null)
```
>if (boundForEachStatement.IterationErrorExpressionOpt != null) [](start = 17, length = 62)

I think we should make more changes to this ```if... else if ... else``` statement. I we have DeconstructionOpt, then loopControlVariable is DeconstructionOpt. If we have IterationErrorExpressionOpt, then loopControlVariable  is IterationErrorExpressionOpt. Only otherwise we should synthesize node for loopControlVariable and I am not even sure if today's final ```else``` should even be reachable.

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IForEachLoopStatement.cs:
```diff
@@ -1279,5 +1280,51 @@ public static void M(int[] x)
+    void M(int a, int b)
+    {
+        int[] arr = new int[10];
+        /*<bind>*/foreach (M(1, 2) in arr)
```
>M(1, 2) [](start = 27, length = 7)

Let's add a tests with an ```out var``` and declaration pattern inside this expression.

AlekseyTs commented:
Done with review pass (iteration 1)

333fred commented:
@AlekseyTs addressed feedback.

---
In reply to: [377820287](https://github.com/dotnet/roslyn/pull/25848#issuecomment-377820287) [](ancestors = 377820287)

333fred commented:
@dotnet/roslyn-compiler I need another review of this please.

333fred commented:
@dotnet-bot retest windows_release_unit32_prtest please

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IForEachLoopStatement.cs:
```diff
@@ -1219,12 +1220,13 @@ public static void M(int[] x)
 IBlockOperation (1 statements) (OperationKind.Block, Type: null, IsInvalid) (Syntax: '{ ... }')
   IForEachLoopOperation (LoopKind.ForEach) (OperationKind.Loop, Type: null, IsInvalid) (Syntax: 'foreach (x[ ... }')
     LoopControlVariable: 
```
>LoopControlVariable [](start = 4, length = 19)

It looks like now IForEachLoopStatement.LoopControlVariable should never be null, please make necessary adjustments (add/change asserts in TestOperationVisitor and OperationTreeVerifier)

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IForEachLoopStatement.cs:
```diff
@@ -1219,12 +1220,13 @@ public static void M(int[] x)
     LoopControlVariable: 
-      null
-    Collection: 
-      IConversionOperation (TryCast: False, Unchecked) (OperationKind.Conversion, Type: System.Collections.IEnumerable, IsImplicit) (Syntax: 'x')
```
>IConversionOperation (TryCast: False, Unchecked) (OperationKind.Conversion, Type: System.Collections.IEnumerable, IsImplicit) (Syntax: 'x') [](start = 6, length = 139)

Do we understand what happened with this conversion?

333fred replied:
Because we set the HasErrors flag, we bail out earlier in ForEachLoopBinder.BindForEachPartsWorker, before this conversion is created.

---
In reply to: [178681842](https://github.com/dotnet/roslyn/pull/25848#discussion_r178681842) [](ancestors = 178681842)

AlekseyTs commented:
https://github.com/dotnet/roslyn/issues/23820 is likely addressed by this change too

333fred commented:
It does appear to be.

---
In reply to: [378085301](https://github.com/dotnet/roslyn/pull/25848#issuecomment-378085301) [](ancestors = 378085301)

333fred commented:
@dotnet/roslyn-compiler I need a second review please.

333fred commented:
@jaredpar for ask mode approval.

jaredpar commented:
approved.
